src/io/sb3/fromSb3.ts

```diff
@@ -423,6 +427,70 @@ export async function fromSb3JSON(json: sb3.ProjectJSON, options: { getAsset: Ge
     videoOn: stage.videoState === "on",
     videoAlpha: stage.videoTransparency
   });
+
+  const targetToBlocks: Map<Target, Array<Block>> = new Map();
+  for (const target of [project.stage, ...project.sprites]) {
+    targetToBlocks.set(target, flattenBlocks(target.scripts.flatMap(script => script.blocks)));
+  }
+  const allBlocks = Array.from(targetToBlocks.values()).flat();
+
+  // Run an extra pass on variables (and lists). Only those which are actually
+  // referenced in blocks or monitors should be kept.
+  for (const target of [project.stage, ...project.sprites]) {
+    let relevantBlocks: Array<Block> = null;
+    if (target === project.stage) {
+      relevantBlocks = allBlocks;
+    } else {
+      relevantBlocks = targetToBlocks.get(target);
+    }
+
+    const usedVariableIds: Set<string> = new Set();
+    for (const block of relevantBlocks) {
+      let id: string = null;
+      if ((block.inputs as { VARIABLE: BlockInput.Variable }).VARIABLE) {
+        id = (block.inputs as { VARIABLE: BlockInput.Variable }).VARIABLE.value.id;
+      } else if ((block.inputs as { LIST: BlockInput.List }).LIST) {
+        id = (block.inputs as { LIST: BlockInput.List }).LIST.value.id;
+      } else {
+        continue;
+      }
+      usedVariableIds.add(id);
+    }
+
+    for (const varList of [target.variables, target.lists]) {
+      for (let i = 0, variable; (variable = varList[i]); i++) {
+        if (variable.visible) {
+          continue;
+        }
+        if (usedVariableIds.has(variable.id)) {
+          continue;
+        }
+
+        varList.splice(i, 1);
+        i--;
+        continue;
+      }
+    }
+  }
+
+  return project;
+}
+
+function flattenBlocks(blocks: Array<Block>) {
+  return blocks.flatMap(block => [
+    block,
+    ...flattenBlocks(
```

Comment:
How many intermediate arrays are we creating here? I'm worried this could be a performance issue for nested blocks, but on the other hand I don't expect the average Scratch project to next blocks that deeply.

Reply:
For instance, could we pass the final block array into `flattenBlocks` and then append to it within each recursive call?